File: apps/glyphshift-desktop/src/workflowLifecycle.ts
import { computed, ref } from 'vue'
import type { WorkflowLifecycle, WorkflowRuntimeStatus } from './model'
import { model } from './workspace/state'

export const workflowOperations = ref<Record<string, 'connecting' | 'stopping' | 'unknown' | undefined>>({})
export function projectLifecycle(runtime: WorkflowRuntimeStatus | undefined, enabled: boolean): WorkflowLifecycle {
  if (runtime?.lifecycle) return runtime.lifecycle
  const active = runtime?.targets.some(target => target.active) ?? false
  const failed = Object.values(runtime?.errors ?? {}).some(error => error.code !== 'runtime.target_not_found')
  return { phase: enabled ? (failed ? 'failed' : active ? 'running' : 'waiting') : (active || failed ? 'stop_failed' : 'stopped'), enabled, collectNewSources: true, checkedAtMs: runtime?.checkedAtMs ?? 0, revision: runtime?.revision ?? 0 }
}
export function workflowLifecycle(id: string): WorkflowLifecycle {
  const state = projectLifecycle(model.value.workflowRuntimeStatus[id], model.value.activations.some(item => item.workflowId === id))
  const operation = workflowOperations.value[id]
  return operation ? { ...state, phase: operation } : state
}
export const workflowActivity = computed(() => {
  const counts: Partial<Record<WorkflowLifecycle['phase'], number>> = {}
  for (const workflow of model.value.workflows) {
    const phase = workflowLifecycle(workflow.id).phase
    if (phase !== 'stopped') counts[phase] = (counts[phase] ?? 0) + 1
  }
  return Object.entries(counts) as [WorkflowLifecycle['phase'], number][]
})

export function workflowStateColor(phase: WorkflowLifecycle['phase']): 'success' | 'error' | 'neutral' | 'warning' {
  return phase === 'running' ? 'success' : ['failed', 'stop_failed'].includes(phase) ? 'error' : phase === 'stopped' ? 'neutral' : 'warning'
}

File: apps/glyphshift-desktop/src-tauri/src/workflow_lifecycle.rs
use super::*;
use std::sync::atomic::{AtomicU64, Ordering};

pub(super) fn next_revision() -> u64 {
    static REVISION: AtomicU64 = AtomicU64::new(1);
    REVISION.fetch_add(1, Ordering::Relaxed)
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub(super) struct WorkflowLifecycle {
    pub(super) phase: &'static str,
    pub(super) enabled: bool,
    pub(super) collect_new_sources: bool,
    pub(super) checked_at_ms: u64,
    pub(super) revision: u64,
}

impl DesktopApplication {
    pub(super) fn project_workflow_runtime(&self, mut runtime: WorkflowRuntimeView) -> WorkflowRuntimeView {
        let enabled = self.backend.enabled_workflow_ids().iter().any(|id| id == &runtime.workflow_id);
        let active = runtime.targets.iter().any(|target| target.active);
        let actionable = runtime.errors.values().any(|error| error.code() != "runtime.target_not_found");
        let phase = if !enabled {
            if active || actionable { "stop_failed" } else { "stopped" }
        } else if actionable { "failed" }
        else if active { "running" }
        else { "waiting" };
        let collect_new_sources = self.backend.workflow(&runtime.workflow_id).ok()
            .is_some_and(|workflow| workflow.targets().iter().any(|target| target.collection_enabled()));
        runtime.lifecycle = Some(WorkflowLifecycle { phase, enabled, collect_new_sources,
            checked_at_ms: runtime.checked_at_ms, revision: runtime.revision });
        runtime
    }
}

// Live sessions still need liveness checks. Only inactive transient failures may
// reconnect automatically; configuration, privilege and resident-module errors wait for the user.
pub(super) fn automatic_retry_allowed(runtime: &WorkflowRuntimeView, now: u64) -> bool {
    if runtime.targets.iter().any(|target| target.active) { return true; }
    let errors = runtime.errors.values().filter(|error| error.code() != "runtime.target_not_found").collect::<Vec<_>>();
    errors.is_empty() || (runtime.retry_attempt < 3 && now >= runtime.retry_after_ms
        && errors.iter().all(|error| error.code() == "runtime.activation_timed_out"))
}
